pallets/confidential-asset/src/lib.rs

```diff
@@ -34,21 +49,112 @@ pub trait Trait:
     + pallet_contracts::Trait
     + pallet_portfolio::Trait
 {
+    type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+}
+
+/// Wrapper for Ciphertexts that correspond to EncryptedAssetId.
+/// This is needed since `mercat::asset_proofs::elgamal_encryption::CipherText` implements
+/// Encode and Decode, instead of deriving them. As a result, the EncodeLike operator is
+/// not automatically implemented.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, VecU8StrongTyped, Default)]
+pub struct EncryptedAssetIdWrapper(pub Vec<u8>);
+
+/// Wrapper for Ciphertexts that correspond to EncryptedBalance.
+/// This is needed since `mercat::asset_proofs::elgamal_encryption::CipherText` implements
+/// Encode and Decode, instead of deriving them. As a result, the EncodeLike operator is
+/// not automatically implemented.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, VecU8StrongTyped, Default)]
+pub struct EncryptedBalanceWrapper(pub Vec<u8>);
+
+/// A mercat account consists of the public key that is used for encryption purposes and the
+/// encrypted asset id. The encrypted asset id also acts as the unique identifier of this
+/// struct.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, Default)]
+pub struct MercatAccount {
+    pub encrypted_asset_id: EncryptedAssetIdWrapper,
+    pub encryption_pub_key: EncryptionPubKey,
 }
 
 decl_storage! {
-    trait Store for Module<T: Trait> as Asset {
+    trait Store for Module<T: Trait> as ConfidentialAsset {
+
+        /// Contains the mercat accounts for an identity.
+        pub MercatAccounts get(fn mercat_account): double_map hasher(twox_64_concat) IdentityId, hasher(blake2_128_concat) EncryptedAssetIdWrapper  => MercatAccount;
+
+        /// Contains the encrypted balance of a mercat account.
+        pub MercatAccountBalance get(fn mercat_account_balance):  double_map hasher(twox_64_concat) IdentityId, hasher(blake2_128_concat) EncryptedAssetIdWrapper => EncryptedBalanceWrapper;
+
+        /// Contains the list of all valid ticker names.
+        /// The process around creating and storing this data will likely change as a result of CRYP-153.
+        pub ValidAssetIds get(fn valid_asset_ids): Vec<AssetId>;
     }
 }
 
 // Public interface for this runtime module.
 decl_module! {
     pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+        /// Initialize the default event for this module
+        fn deposit_event() = default;
+
+        /// TODO: CRYP-153 will most likely change this. The modified version of this function
+        /// will be called when a new ticker name is registered.
+        #[weight = 1_000_000_000]
+        pub fn create_confidential_asset(
+            origin,
+            valid_asset_ids: Vec<AssetId>,
+        ) -> DispatchResult
+        {
```

Comment:
A very minor point but `rustfmt` would put `{` at the end of the previous line, `) -> DispatchResult {`